packages/cursorless-engine/src/core/commandVersionUpgrades/upgradeV5ToV6/upgradeV5ToV6.ts

```diff
@@ -0,0 +1,359 @@
+      };
+    case "replace":
+      return {
+        name,
+        replaceWith: action.args![0] as ReplaceWith,
+        destination: targetToDestination(targets[0]),
+      };
+    case "highlight":
+      return {
+        name,
+        highlightId: action.args?.[0] as HighlightId | undefined,
+        target: upgradeTarget(targets[0]),
+      };
+    case "editNew":
+      return {
+        name,
+        destination: targetToDestination(targets[0]),
+      };
+    default:
+      return {
+        name,
+        target: upgradeTarget(targets[0]),
+      };
+  }
+}
+
+function upgradeTarget(
+  target: PartialTargetDescriptorV5,
+): PartialTargetDescriptor {
+  switch (target.type) {
+    case "list":
+    case "range":
+    case "primitive":
+      return upgradeNonImplicitTarget(target);
+    case "implicit":
+      return target;
+  }
+}
+
+function upgradeNonImplicitTarget(
+  target:
+    | PartialPrimitiveTargetDescriptorV5
+    | PartialRangeTargetDescriptorV5
+    | PartialListTargetDescriptorV5,
+) {
+  switch (target.type) {
+    case "list":
+      return upgradeListTarget(target);
+    case "range":
+    case "primitive":
+      return upgradeRangeOrPrimitiveTarget(target);
+  }
+}
+
+function upgradeListTarget(
+  target: PartialListTargetDescriptorV5,
+): PartialListTargetDescriptor {
+  return {
+    ...target,
+    elements: target.elements.map(upgradeRangeOrPrimitiveTarget),
+  };
+}
+
+function upgradeRangeOrPrimitiveTarget(
+  target: PartialPrimitiveTargetDescriptorV5 | PartialRangeTargetDescriptorV5,
+) {
+  switch (target.type) {
+    case "range":
+      return upgradeRangeTarget(target);
+    case "primitive":
+      return upgradePrimitiveTarget(target);
+  }
+}
+
+function upgradeRangeTarget(
+  target: PartialRangeTargetDescriptorV5,
+): PartialRangeTargetDescriptor {
+  const { anchor, active } = target;
+  return {
+    ...target,
+    anchor:
+      anchor.type === "implicit" ? anchor : upgradePrimitiveTarget(anchor),
+    active: upgradePrimitiveTarget(active),
+  };
+}
+
+function upgradePrimitiveTarget(
+  target: PartialPrimitiveTargetDescriptorV5,
+): PartialPrimitiveTargetDescriptor {
+  return {
+    ...target,
+    modifiers: upgradeModifiers(target.modifiers),
+  };
+}
+
+function targetToDestination(
+  target: PartialTargetDescriptorV5,
+): DestinationDescriptor {
+  switch (target.type) {
+    case "list":
+      return listTargetToDestination(target);
+    case "range":
+      return rangeTargetToDestination(target);
+    case "primitive":
+      return primitiveTargetToDestination(target);
+    case "implicit":
+      return implicitTargetToDestination();
+  }
+}
+
+/**
+ * Converts a list target to a destination. This is a bit tricky because we need
+ * to split the list into multiple destinations if there is more than one insertion
+ * mode.
+ * @param target The target to convert
+ * @returns The converted destination
+ */
+function listTargetToDestination(
```

Comment:
Your implementation was creating one destination per element of a list target.  E.g. it was mapping `"to end of air and end of bat"` to `"to end of air and to end of bat"` (note the second `"to"`).  I changed it so it only creates a list of there are multiple insertion modes, which is what the Talon side will do in practice.  For the most part, there is no difference, except for one small fact: if the targets refer to the same range, they get deduplicated in my implementation, because they will go through the target pipeline together, whereas in your version they would not.  This actually affected a couple unit tests:

- It turns out I had to roll back 5a95c6517d27e8b590285986c7c9cea3f4e31c8f because that did actually have an effect, it's just that without deduplication the insertion delimiter joining wasn't actually being exercised for a couple tests because they weren't getting deduplicated into a single target
- I was able to revert your change to `bringPointAndHarpToEndOfThisAndEndOfWhaleTakeWhale`.  The lack of deduplication was what was causing your test case failure there.

I wonder if we should have a final deduplication step for destinations where we deduplicate if they are the same target and insertion mode 🤔.  I would be tempted to handle that in follow-up work, but wouldn't be too bad to tackle in this PR; wdyt?